3rd attempt at fixing 42957. Replaces PRs 1095, 1896

daveagp opened the pull request:
On my website we get about one email a month from a person who can't create an account. This turns out to be caused by https://core.trac.wordpress.org/ticket/42957

- you're using gmail
- you try to create an account name ending in a period, e.g. Dave P.
- the email is sent containing [https://blah...&login=Dave%20P](https://blah...%26login%3Ddave%20p/).
- gmail generally assumes links ending in periods have the periods added by humans and thus ignores them
- so clicking on it, you arrive at [https://blah...&login=Dave%20P](https://blah...%26login%3Ddave%20p/) and the user's told the link is invalid

Fixes this by encoding the trailing period as an entity.

This is my first submitted pull request to WP (technically second but it's a fix to the first), thanks in advance for telling me if any more steps are needed.

Trac ticket: https://core.trac.wordpress.org/ticket/42957

daveagp committed:
copy of https://github.com/WordPress/wordpress-develop/pull/1896/files

daveagp committed:
suggested rename

daveagp committed:
missed a rename

daveagp committed:
fix renames. unit tests pass

daveagp committed:
add missing trailing newline caught by php coding standards

mukeshpanchal27 commented on src/wp-includes/functions.php:
```diff
@@ -8860,6 +8860,39 @@ function wp_fuzzy_number_match( $expected, $actual, $precision = 1 ) {
+ * @param string $url The URL to encode.
+ * @return string The new encoded URL.
+ */
+function _encode_query_arg_ending_period( $url ) {
```
Does this function private?

daveagp replied:
I think I found out the way to do it is @access private? I've made a change to do that now

mukeshpanchal27 commented on tests/phpunit/tests/functions/encodeQueryArgEndingPeriod.php:
```diff
@@ -0,0 +1,84 @@
+class Tests_Functions__EncodeQueryArgEndingPeriod extends WP_UnitTestCase {
+
+	/**
+	 * @dataProvider data__encode_query_arg_ending_period
+	 *
+	 * @ticket 42957
```
```suggestion
	 * @ticket 42957
	 *
	 * @dataProvider data__encode_query_arg_ending_period
```

mukeshpanchal27 commented on src/wp-includes/functions.php:
```diff
@@ -8860,6 +8860,39 @@ function wp_fuzzy_number_match( $expected, $actual, $precision = 1 ) {
+ * If the URL ends with a period, replace it with %2E to avoid
+ * misinterpretation from some email clients.
+ *
+ * @since 5.9.0
```
```suggestion
 * @since 6.6.0
```

daveagp committed:
Update tests/phpunit/tests/functions/encodeQueryArgEndingPeriod.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

daveagp committed:
Update src/wp-includes/functions.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

daveagp committed:
Add @access private, following example of other functions

dmsnell commented on src/wp-includes/functions.php:
```diff
@@ -8860,6 +8860,40 @@ function wp_fuzzy_number_match( $expected, $actual, $precision = 1 ) {
+		return '';
+	}
+
+	$url = rawurlencode( $url );
```
note that we don't want to call `rawurlencode()` on a full URL. it's only meant to encode the parts after the domain, and we want to avoid double-encoding it, since if we call this on an already-escaped URL, we'll end up escaping the escapes and corrupting it.

daveagp replied:
Fair enough, though I am going to update my pull request to match your other very elegant solution

daveagp committed:
Delete src/wp-includes/functions.php

Step 1/4 to implement cleaner solution due to dmnell

https://core.trac.wordpress.org/ticket/42957#comment:37

daveagp committed:
revert file

daveagp committed:
Delete tests/phpunit/tests/functions/encodeQueryArgEndingPeriod.php

Step 2/4

daveagp committed:
Update pluggable.php

First edit with nice solution

daveagp committed:
Update user.php - final elegant dmsnell fix 

See https://core.trac.wordpress.org/ticket/42957#comment:37

daveagp committed:
Revert changes in functions.php

daveagp commented:
I've reimplemented using the nice solution from https://core.trac.wordpress.org/ticket/42957#comment:37. Thanks @dmsnell for the review and thoughtful suggestion!

daveagp committed:
Update pluggable.php - less fancy quotes

daveagp committed:
Update user.php - less fancy quotes

dmsnell reviewed (approved):
> I've reimplemented…Thanks @dmsnell for the review and thoughtful suggestion!

Happy to help. Any problem we _don't have to solve_ is usually better in my opinion than any solution to that problem 😉

daveagp committed:
Update pluggable.php - add comment

daveagp committed:
Update user.php - add comment

dmsnell committed:
Merge branch 'trunk' into 42957b

dmsnell committed:
Expand comment warning not to move the query arg.

dmsnell commented:
Merged in [[58674]](https://core.trac.wordpress.org/changeset/58674)
https://github.com/wordpress/wordpress-develop/commit/82deed39a56f46bccde20724065ef0d50455f3a7
